src/shared/particle_dynamics/continuum_dynamics/continuum_integration.cpp

```diff
@@ -116,14 +125,117 @@ void StressDiffusion::interaction(size_t index_i, Real dt)
         Real dW_ijV_j = inner_neighborhood.dW_ij_[n] * Vol_[index_j];
         Real y_ij = pos_[index_i](1, 0) - pos_[index_j](1, 0);
         diffusion_stress_ = stress_tensor_3D_[index_i] - stress_tensor_3D_[index_j];
-        diffusion_stress_(0, 0) -= (1 - sin(fai_)) * density * gravity * y_ij;
+        diffusion_stress_(0, 0) -= (1 - sin(phi_)) * density * gravity * y_ij;
         diffusion_stress_(1, 1) -= density * gravity * y_ij;
-        diffusion_stress_(2, 2) -= (1 - sin(fai_)) * density * gravity * y_ij;
+        diffusion_stress_(2, 2) -= (1 - sin(phi_)) * density * gravity * y_ij;
         diffusion_stress_rate_ += 2 * zeta_ * smoothing_length_ * sound_speed_ *
                                   diffusion_stress_ * r_ij * dW_ijV_j / (r_ij * r_ij + 0.01 * smoothing_length_);
     }
     stress_rate_3D_[index_i] = diffusion_stress_rate_;
 }
 //====================================================================================//
+ShearStressRelaxationHourglassControl1stHalf ::
+    ShearStressRelaxationHourglassControl1stHalf(BaseInnerRelation &inner_relation, Real xi)
+    : fluid_dynamics::BaseIntegration<DataDelegateInner>(inner_relation),
+      continuum_(DynamicCast<GeneralContinuum>(this, particles_->getBaseMaterial())),
+      shear_stress_(particles_->registerStateVariable<Matd>("ShearStress")),
+      shear_stress_rate_(particles_->registerStateVariable<Matd>("ShearStressRate")),
+      velocity_gradient_(particles_->registerStateVariable<Matd>("VelocityGradient")),
+      strain_tensor_(particles_->registerStateVariable<Matd>("StrainTensor")),
+      strain_tensor_rate_(particles_->registerStateVariable<Matd>("StrainTensorRate")),
+      B_(particles_->getVariableDataByName<Matd>("LinearGradientCorrectionMatrix")),
+      scale_penalty_force_(particles_->registerStateVariable<Real>("ScalePenaltyForce")), xi_(xi)
+{
+    particles_->addVariableToSort<Matd>("ShearStress");
+    particles_->addVariableToSort<Matd>("ShearStressRate");
+    particles_->addVariableToSort<Matd>("VelocityGradient");
+    particles_->addVariableToSort<Matd>("StrainTensor");
+    particles_->addVariableToSort<Matd>("StrainTensorRate");
+    particles_->addVariableToSort<Real>("ScalePenaltyForce");
```

Comment:
Please make sure that all these variable need to be remembered during the simulation. For example, it seems VelocityGradient is only temporary variable refreshed every time step.

Reply:
`VelocityGradient ` is used at different places. `ShearStressRate `and `StrainTensorRate `are locally and don't need to be registered. I will delete them.

Reply:
This is done.